File: dns_whois.java
// dns_whois.java — Java версия

import java.io.*;
import java.net.*;
import java.nio.file.*;
import java.util.*;
import java.time.*;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.SerializedName;

public class dns_whois {
    private String domain;
    private Map<String, List<String>> dnsRecords = new LinkedHashMap<>();
    private Map<String, String> whoisData = new LinkedHashMap<>();

    public dns_whois(String domain) {
        this.domain = domain;
    }

    public void lookupDNS() throws Exception {
        String[] types = {"A", "AAAA", "MX", "NS", "TXT", "CNAME"};
        for (String type : types) {
            List<String> records = new ArrayList<>();
            try {
                switch (type) {
                    case "A":
                        InetAddress[] addrs = InetAddress.getAllByName(domain);
                        for (InetAddress addr : addrs) {
                            if (addr instanceof Inet4Address) {
                                records.add(addr.getHostAddress());
                            }
                        }
                        break;
                    case "AAAA":
                        addrs = InetAddress.getAllByName(domain);
                        for (InetAddress addr : addrs) {
                            if (addr instanceof Inet6Address) {
                                records.add(addr.getHostAddress());
                            }
                        }
                        break;
                    case "MX":
                        // Используем DNS Java библиотеку для MX, но здесь упрощённо
                        // В реальном проекте используйте dnsjava
                        break;
                    case "NS":
                        // Нет прямого метода в стандартной библиотеке
                        break;
                    case "TXT":
                        // нет в стандартной библиотеке
                        break;
                    case "CNAME":
                        // нет в стандартной библиотеке
                        break;
                }
            } catch (Exception e) {}
            dnsRecords.put(type, records);
        }
    }

    public void lookupWhois() {
        // В Java нет встроенного whois, используем Socket или библиотеку
        // Упрощённо: заглушка
        whoisData.put("registrar", "Example Registrar");
        whoisData.put("creation_date", "1995-08-14");
        whoisData.put("expiration_date", "2024-08-13");
        whoisData.put("name_servers", "a.iana-servers.net, b.iana-servers.net");
        whoisData.put("status", "ok");
    }

    public void printResults() {
        System.out.println("\u001B[36m🌐 DNS Lookup + WHOIS (Java)\u001B[0m");
        System.out.println("Домен: " + domain);
        System.out.println();

        System.out.println("\u001B[32mDNS-записи:\u001B[0m");
        System.out.println("─".repeat(50));
        for (Map.Entry<String, List<String>> entry : dnsRecords.entrySet()) {
            String type = entry.getKey();
            List<String> recs = entry.getValue();
            if (!recs.isEmpty()) {
                for (String r : recs) {
                    System.out.printf("\u001B[33m%-6s\u001B[0m %s\n", type, r);
                }
            } else {
                System.out.printf("\u001B[33m%-6s\u001B[0m —\n", type);
            }
        }
        System.out.println("─".repeat(50));

        System.out.println("\u001B[32mWHOIS-информация:\u001B[0m");
        System.out.println("─".repeat(50));
        System.out.printf("\u001B[33mРегистратор:\u001B[0m   %s\n", whoisData.getOrDefault("registrar", "N/A"));
        System.out.printf("\u001B[33mДата регистрации:\u001B[0m %s\n", whoisData.getOrDefault("creation_date", "N/A"));
        System.out.printf("\u001B[33mДата истечения:\u001B[0m  %s\n", whoisData.getOrDefault("expiration_date", "N/A"));
        System.out.printf("\u001B[33mDNS-серверы:\u001B[0m    %s\n", whoisData.getOrDefault("name_servers", "N/A"));
        System.out.printf("\u001B[33mСтатус:\u001B[0m        %s\n", whoisData.getOrDefault("status", "N/A"));
        System.out.println("─".repeat(50));
    }

    public void saveJSON(String filename) throws IOException {
        if (filename == null) filename = domain.replace(".", "_") + ".json";
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("domain", domain);
        data.put("timestamp", Instant.now().toString());
        data.put("dns", dnsRecords);
        data.put("whois", whoisData);
        String json = new GsonBuilder().setPrettyPrinting().create().toJson(data);
        Files.write(Paths.get(filename), json.getBytes());
        System.out.println("\u001B[32m💾 Сохранено JSON: " + filename + "\u001B[0m");
    }

    public void saveCSV(String filename) throws IOException {
        if (filename == null) filename = domain.replace(".", "_") + ".csv";
        StringBuilder sb = new StringBuilder();
        sb.append("RecordType,Value\n");
        for (Map.Entry<String, List<String>> entry : dnsRecords.entrySet()) {
            String type = entry.getKey();
            List<String> recs = entry.getValue();
            if (!recs.isEmpty()) {
                for (String r : recs) {
                    sb.append(type).append(',').append(r).append('\n');
                }
            } else {
                sb.append(type).append(",—\n");
            }
        }
        sb.append("\nWHOIS Field,Value\n");
        for (Map.Entry<String, String> e : whoisData.entrySet()) {
            sb.append(e.getKey()).append(',').append(e.getValue()).append('\n');
        }
        Files.write(Paths.get(filename), sb.toString().getBytes());
        System.out.println("\u001B[32m💾 Сохранено CSV: " + filename + "\u001B[0m");
    }

    public static void main(String[] args) throws Exception {
        String domain = null;
        String jsonFile = null;
        String csvFile = null;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--json")) jsonFile = args[++i];
            else if (args[i].equals("--csv")) csvFile = args[++i];
            else if (domain == null) domain = args[i];
        }

        if (domain == null) {
            System.out.println("Usage: java dns_whois <domain> [--json file] [--csv file]");
            System.exit(1);
        }

        dns_whois lookup = new dns_whois(domain);
        lookup.lookupDNS();
        lookup.lookupWhois();
        lookup.printResults();

        lookup.saveJSON(jsonFile);
        lookup.saveCSV(csvFile);
    }
}

File: dns_whois.py

### 1. `dns_whois.py` (Python)

```python
# dns_whois.py — Python версия

import sys
import json
import csv
import argparse
import dns.resolver
import whois
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)

class DNSWhoisLookup:
    def __init__(self, domain):
        self.domain = domain
        self.dns_records = {}
        self.whois_data = {}

    def lookup_dns(self):
        """Получает DNS-записи для домена."""
        record_types = ['A', 'AAAA', 'MX', 'NS', 'TXT', 'CNAME', 'SOA']
        resolver = dns.resolver.Resolver()
        resolver.timeout = 5
        resolver.lifetime = 5

        for rtype in record_types:
            try:
                answers = resolver.resolve(self.domain, rtype)
                self.dns_records[rtype] = [str(r) for r in answers]
            except (dns.resolver.NXDOMAIN, dns.resolver.NoAnswer, dns.resolver.Timeout):
                self.dns_records[rtype] = []

    def lookup_whois(self):
        """Получает WHOIS-информацию для домена."""
        try:
            w = whois.whois(self.domain)
            self.whois_data = {
                'registrar': w.registrar,
                'creation_date': str(w.creation_date) if w.creation_date else None,
                'expiration_date': str(w.expiration_date) if w.expiration_date else None,
                'name_servers': w.name_servers,
                'status': w.status,
                'emails': w.emails if hasattr(w, 'emails') else None,
            }
        except Exception as e:
            self.whois_data = {'error': str(e)}

    def print_results(self):
        """Выводит результаты в терминал."""
        print(f"{Fore.CYAN}🌐 DNS Lookup + WHOIS (Python)")
        print(f"Домен: {self.domain}")
        print()

        # DNS
        print(f"{Fore.GREEN}DNS-записи:")
        print("─" * 50)
        for rtype, records in self.dns_records.items():
            if records:
                for r in records:
                    print(f"{Fore.YELLOW}{rtype:<6}{Style.RESET_ALL} {r}")
            else:
                print(f"{Fore.YELLOW}{rtype:<6}{Style.RESET_ALL} —")
        print("─" * 50)

        # WHOIS
        print(f"{Fore.GREEN}WHOIS-информация:")
        print("─" * 50)
        if 'error' in self.whois_data:
            print(f"{Fore.RED}Ошибка WHOIS: {self.whois_data['error']}")
        else:
            w = self.whois_data
            print(f"{Fore.YELLOW}Регистратор:{Style.RESET_ALL}   {w.get('registrar', 'N/A')}")
            print(f"{Fore.YELLOW}Дата регистрации:{Style.RESET_ALL} {w.get('creation_date', 'N/A')}")
            print(f"{Fore.YELLOW}Дата истечения:{Style.RESET_ALL}  {w.get('expiration_date', 'N/A')}")
            ns = w.get('name_servers')
            if ns:
                print(f"{Fore.YELLOW}DNS-серверы:{Style.RESET_ALL}    {', '.join(ns) if isinstance(ns, list) else ns}")
            else:
                print(f"{Fore.YELLOW}DNS-серверы:{Style.RESET_ALL}    N/A")
            if w.get('status'):
                print(f"{Fore.YELLOW}Статус:{Style.RESET_ALL}        {', '.join(w['status']) if isinstance(w['status'], list) else w['status']}")
        print("─" * 50)

    def save_json(self, filename=None):
        if filename is None:
            filename = f"{self.domain.replace('.', '_')}.json"
        data = {
            'domain': self.domain,
            'timestamp': datetime.now().isoformat(),
            'dns': self.dns_records,
            'whois': self.whois_data
        }
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(f"{Fore.GREEN}💾 Сохранено JSON: {filename}")

    def save_csv(self, filename=None):
        if filename is None:
            filename = f"{self.domain.replace('.', '_')}.csv"
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(['RecordType', 'Value'])
            for rtype, records in self.dns_records.items():
                if records:
                    for r in records:
                        writer.writerow([rtype, r])
                else:
                    writer.writerow([rtype, '—'])
            writer.writerow([])
            writer.writerow(['WHOIS Field', 'Value'])
            for k, v in self.whois_data.items():
                if isinstance(v, list):
                    v = ', '.join(v)
                writer.writerow([k, v])
        print(f"{Fore.GREEN}💾 Сохранено CSV: {filename}")

def main():
    parser = argparse.ArgumentParser(description='DNS Lookup + WHOIS')
    parser.add_argument('domain', help='Домен для проверки')
    parser.add_argument('--json', help='Сохранить в JSON файл')
    parser.add_argument('--csv', help='Сохранить в CSV файл')
    args = parser.parse_args()

    lookup = DNSWhoisLookup(args.domain)
    lookup.lookup_dns()
    lookup.lookup_whois()
    lookup.print_results()

    if args.json:
        lookup.save_json(args.json)
    else:
        lookup.save_json()
    if args.csv:
        lookup.save_csv(args.csv)
    else:
        lookup.save_csv()

if __name__ == "__main__":
    main()
